src/push_context.rs

```diff
@@ -0,0 +1,397 @@
+use std::{
+    collections::HashSet,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
+use color_eyre::eyre::{eyre, Context, Result};
+
+use crate::{
+    build_http_client,
+    cli::FlakeHubPushCli,
+    flake_info, flakehub_auth_fake,
+    flakehub_client::Tarball,
+    git_context::GitContext,
+    github::graphql::{GithubGraphqlDataQuery, MAX_LABEL_LENGTH, MAX_NUM_TOTAL_LABELS},
+    release_metadata::ReleaseMetadata,
+    revision_info::RevisionInfo,
+    DEFAULT_ROLLING_PREFIX,
+};
+
+#[derive(Clone)]
+pub enum ExecutionEnvironment {
+    GitHub,
+    GitLab,
+    Local,
+}
+
+pub(crate) struct PushContext {
+    pub(crate) flakehub_host: url::Url,
+    pub(crate) auth_token: String,
+
+    // url components
+    pub(crate) upload_name: String, // {org}/{project}
+    pub(crate) release_version: String,
+
+    // internal behavior changes
+    pub(crate) error_if_release_conflicts: bool,
+
+    // the goods
+    pub(crate) metadata: ReleaseMetadata,
+    pub(crate) tarball: Tarball,
+}
+
+impl PushContext {
+    pub async fn from_cli_and_env(cli: &mut FlakeHubPushCli) -> Result<Self> {
+        // Take the opportunity to be able to populate/encrich data from the GitHub API
+        // this is used to augment user/discovered data, and is used for the faked JWT for local flakehub-push testing
+
+        let client = build_http_client().build()?;
+
+        let exec_env = if std::env::var("GITHUB_ACTION").ok().is_some() {
+            ExecutionEnvironment::GitHub
+        } else if std::env::var("GITLAB_CI").ok().is_some() {
+            ExecutionEnvironment::GitLab
+        } else {
+            ExecutionEnvironment::Local
+        };
+
+        match exec_env.clone() {
+            ExecutionEnvironment::GitHub => {
+                // TODO(colemickens): we were back-filling from github, in local paths before, check
```

Comment:
?

Reply:
it's fine, easy to add back if we decide we really want/need it for local dev